lib/tree.rb

```diff
@@ -1,64 +1,179 @@
+        current = current.left
+      elsif key > current.key
+        current = current.right
+      end
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n) - n is the number of nodes in the tree
+  # Space Complexity: O(n) - my thought is that since we're making a list of nodes to return, we'll have that many nodes stored in the list, but since the stack doesn't ever hold something that's also already in the list, there would never be more than n nodes stored at one time..?  Let me know if that's on the right track.
+  # left, root, right (me btwn my children)
   def inorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    list = []
+    stack = []
+    current = @root
+    
+    # while current node exists OR the stack is not empty
+    while !current.nil? || !stack.empty?
+      
+      # while current node exists
+      while !current.nil?
+        # add it to the stack
+        stack << current
+        # current becomes node to its left and if it is not nil, this loop will continue
+        current = current.left
+      end
+      
+      # pop off the last node that was added
+      current = stack.pop
+      # add it to the list
+      list << {key: current.key, value: current.value}
+      # traverse to the right
+      current = current.right
+    end
+    
+    return list
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
-  def preorder
-    raise NotImplementedError
+  
+  # Time Complexity: O(n) - n is the number of nodes in the tree
+  # Space Complexity: O(n) - same assumption as inorder
+  # root, left, right (me before my children)
+  def preorder 
+    return [] if @root.nil?
+    list = []
+    stack = []
+    current = @root
+    
+    stack.push(current)
+    while !stack.empty?
+      current = stack.pop
+      list << {key: current.key, value: current.value}
+      stack << current.right if current.right
+      stack << current.left if current.left
+    end
+    
+    return list
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
-  def postorder
-    raise NotImplementedError
+  
+  # Time Complexity: O(n) - n is the number of nodes in the tree
+  # Space Complexity: O(n) - same assumption as inorder
+  # left, right, root (me after my children)
+  def postorder 
+    return [] if @root.nil?
+    
+    list = []
+    s1 = []
+    s2 = []
+    current = @root
+    
+    s1 << current
+    while !s1.empty?
+      current = s1.pop
+      s2 << current
+      s1 << current.left if current.left
+      s1 << current.right if current.right
+    end
+    
+    while !s2.empty?
+      current = s2.pop
+      list << {key: current.key, value: current.value}
+    end
+    
+    return list
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: Olog(n) in the best case (e.g. if the tree is balanced), O(n) in the worst case (the tree is unbalanced and essentially a linked list)
+  # Space Complexity: O(1)
   def height
```

Comment:
👍 , Since you are doing this recursively the space complexity is O(log n) if the tree is balanced and O(n) otherwise.